File: apps/web/test/music_service_contract.test.js
const assert = require("node:assert/strict");
const { readFileSync } = require("node:fs");
const path = require("node:path");
const test = require("node:test");

const contractPath = path.resolve(__dirname, "../../../tests/fixtures/music_service_contract.json");
const contract = JSON.parse(readFileSync(contractPath, "utf8"));
const workerModule = import("../worker/_worker.js");

function evaluate(worker, operation, input) {
  const common = {
    language: input.language,
    preferFlat: input.prefer_flat,
  };
  if (operation === "generate_chord") {
    return worker.generateChord({
      ...common,
      rootPc: input.root_pc,
      suffix: input.suffix,
      inversion: input.inversion,
    });
  }
  if (operation === "generate_scale") {
    return worker.generateScale({
      ...common,
      tonicPc: input.tonic_pc,
      patternName: input.pattern_name,
    });
  }
  if (operation === "detect_chord") {
    return worker.detectChord({ ...common, notes: input.notes });
  }
  throw new Error(`Unknown contract operation: ${operation}`);
}

for (const contractCase of contract.cases) {
  test(`worker music contract: ${contractCase.id}`, async () => {
    const worker = await workerModule;
    const result = evaluate(worker, contractCase.operation, contractCase.input);
    const actual = Object.fromEntries(
      Object.keys(contractCase.expected).map((field) => [field, result[field]]),
    );
    assert.deepEqual(actual, contractCase.expected);
  });
}

File: apps/mobile_flutter/test/music_service_contract_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:midichords/music_service.dart';

Map<String, dynamic> evaluateContractCase(
  String operation,
  Map<String, dynamic> input,
) {
  if (operation == 'generate_chord') {
    return generateChordLocal(
      rootPc: input['root_pc'] as int,
      suffix: input['suffix'] as String,
      inversion: input['inversion'] as int,
      language: input['language'] as String,
      preferFlat: input['prefer_flat'] as bool,
    );
  }
  if (operation == 'generate_scale') {
    return generateScaleLocal(
      tonicPc: input['tonic_pc'] as int,
      patternName: input['pattern_name'] as String,
      language: input['language'] as String,
      preferFlat: input['prefer_flat'] as bool,
    );
  }
  if (operation == 'detect_chord') {
    return detectChordLocal(
      notes: (input['notes'] as List<dynamic>).cast<int>(),
      language: input['language'] as String,
      preferFlat: input['prefer_flat'] as bool,
    );
  }
  throw StateError('Unknown contract operation: $operation');
}

void main() {
  final contractFile = File('../../tests/fixtures/music_service_contract.json');
  final contract =
      jsonDecode(contractFile.readAsStringSync()) as Map<String, dynamic>;
  final cases = (contract['cases'] as List<dynamic>)
      .cast<Map<String, dynamic>>();

  for (final contractCase in cases) {
    test('mobile music contract: ${contractCase['id']}', () {
      final expected = contractCase['expected'] as Map<String, dynamic>;
      final result = evaluateContractCase(
        contractCase['operation'] as String,
        contractCase['input'] as Map<String, dynamic>,
      );
      final actual = <String, dynamic>{
        for (final field in expected.keys) field: result[field],
      };
      expect(actual, expected);
    });
  }
}
